corehq/apps/commtrack/tests/test_rebuild.py

```diff
@@ -0,0 +1,85 @@
+from django.test import TestCase
+from casexml.apps.case.models import CommCareCase
+from casexml.apps.stock.models import StockTransaction
+from corehq.apps.commtrack.helpers import make_product
+from corehq.apps.commtrack.models import StockState
+from corehq.apps.commtrack.processing import rebuild_stock_state
+from corehq.apps.hqcase.utils import submit_case_blocks
+
+
+LEDGER_BLOCKS_SIMPLE = """
+<transfer xmlns="http://commcarehq.org/ledger/v1" dest="{case_id}" date="2000-01-02" section-id="stock">
+    <entry id="{product_id}" quantity="100" />
+</transfer >
+
+<balance xmlns="http://commcarehq.org/ledger/v1" entity-id="{case_id}" date="2000-01-01" section-id="stock">
+    <entry id="{product_id}" quantity="100" />
+</balance>
+"""
+
+LEDGER_BLOCKS_INFERRED = """
+<transfer xmlns="http://commcarehq.org/ledger/v1" dest="{case_id}" date="2000-01-02" section-id="stock">
+    <entry id="{product_id}" quantity="50" />
+</transfer >
+
+<balance xmlns="http://commcarehq.org/ledger/v1" entity-id="{case_id}" date="2000-01-01" section-id="stock">
+    <entry id="{product_id}" quantity="100" />
+</balance>
+"""
+
+
+class RebuildStockStateTest(TestCase):
+    def setUp(self):
+        self.domain = 'asldkjf-domain'
+        self.case = CommCareCase(domain=self.domain)
+        self.case.save()
+        self.product = make_product(self.domain, 'Product Name', 'prodcode')
+        self._stock_state_key = dict(
+            section_id='stock',
+            case_id=self.case.get_id,
+            product_id=self.product.get_id
+        )
+
+    def _get_stats(self):
+        stock_state = StockState.objects.get(**self._stock_state_key)
+        latest_txn = StockTransaction.latest(**self._stock_state_key)
+        all_txns = StockTransaction.get_ordered_transactions_for_stock(
+            **self._stock_state_key)
+        return stock_state, latest_txn, all_txns
+
+    def _submit_ledgers(self, ledger_blocks):
+        submit_case_blocks(ledger_blocks.format(**self._stock_state_key),
+                           self.domain)
+
+    def test_simple(self):
+        self._submit_ledgers(LEDGER_BLOCKS_SIMPLE)
+        stock_state, latest_txn, all_txns = self._get_stats()
+        self.assertEqual(stock_state.stock_on_hand, 100)
+        self.assertEqual(latest_txn.stock_on_hand, 100)
+        self.assertEqual(all_txns.count(), 2)
+
+        rebuild_stock_state(**self._stock_state_key)
+
+        stock_state, latest_txn, all_txns = self._get_stats()
+        self.assertEqual(stock_state.stock_on_hand, 200)
+        self.assertEqual(latest_txn.stock_on_hand, 200)
+        self.assertEqual(all_txns.count(), 2)
```

Comment:
You probably saw this, but it "makes sense" (slash, this is the behavior I was trying to induce) because `LEDGER_BLOCKS_SIMPLE` contains out-of-date-order blocks, which are then rectified by `rebuild_stock_state`. (Again, once/if/when we decide to sort by server timestamp instead, I'm going to have to rewrite this test.)

Does that make sense? Are you good with this test, and just unsettled by the implications, or do you think I should be looking harder at something?